src/libs/API.js

```diff
@@ -176,22 +169,28 @@ function handleExpiredAuthToken(originalResponse, originalCommand, originalParam
  * @returns {Promise}
  */
 function request(command, parameters, type = 'post') {
-    const networkPromise = Network.post(command, parameters, type);
-
-    // Setup the default handlers to work with different response codes
-    networkPromise.then((response) => {
-        // Handle expired auth tokens properly
-        if (response.jsonCode === 407) {
-            handleExpiredAuthToken(response, command, parameters, type);
-
-            // Throw an error to prevent other handlers from being triggered on this promise
-            throw new Error('A default handler was used for this request');
-        }
-
-        return response;
+    return new Promise((resolve, reject) => {
+        Network.post(command, parameters, type)
+            .then((response) => {
+                // Handle expired auth tokens properly by making sure to pass the resolve and reject down to the
+                // new promise created when calling handleExpiredAuthToken.
+                if (response.jsonCode === 407) {
+                    // There are some API requests that should not be retried when there is an auth failure like
+                    // creating and deleting logins
+                    if (parameters.doNotRetry) {
+                        return;
```

Comment:
Is it OK that this results in an unresolved promise? I think so, but I wanted to double-check since it's not quite obvious that would be what happens. I think the only alternative in this case would be to reject the promise... There are a few API commands that use `doNotRetry` and also have `catch()`. In some cases the error is displayed in the sign in UI. In another case the error is just logged to the console. 

However, in the Pusher authentication case, it triggers a reconnection of Pusher, which is maybe not what to have happen, so an unresolved promise might still be best here.

Reply:
Great thought and I had the same question and waffled back and forth on the best way to handle this. I research a bit and came upon [this SO post](https://stackoverflow.com/questions/36734900/what-happens-if-you-dont-resolve-or-reject-a-promise#:~:text=A%20promise%20is%20just%20an,just%20a%20regular%20Javascript%20object.). It seems like an unresolved Promise is OK because a Promise is "just a JavaScript object". Whether it resolves or not perhaps isn't a big deal unless you need to _know_ that it has resolved or rejected for some other reason. Thinking on this some more and I'm not 100% sure if those promises would be garbage collected or not (but I'd hope maybe they would).

I think this is important to understand and think about what was happening before these changes are applied so I'm gonna ruber duck for a second. 

**Before:** 
- If we encounter a `doNotRetry` command then that original `Promise` still resolves with the `jsonCode: 407` in the `.then()` of whatever `API.Command` calls it (since `HttpUtils.xhr()` is where we'd resolve that original `Promise`).

**After these changes:**
- Instead of returning the error or response from `fetch()` back to the original caller and fulfilling that `Promise` we just won't resolve the `Promise` at all if we have a `doNotRetry`.

It looks like there are some commands that have `doNotRetry` and `catch()` to handle any errors. So if we want to make the safest change possible we could resolve or reject the initial response from `fetch()` back to `Promise` - and worry about whether this makes sense in the future.

But maybe we can just audit all these usages now and figure out why they are `doNotRetry` commands.

`Push_Authenticate` - Looks like maybe it _should_ be retried but currently has it's own handling of the `407` which maybe can be cleaned up if we remove the `doNotRetry`.

https://github.com/Expensify/Expensify.cash/blob/74d33dfde6aadb07456a6ce750836528f70f80e5/src/libs/PusherConnectionManager.js#L26-L41

`DeleteLogin` looks like maybe we just want to return any caught errors, but don't care about the response itself.

`Authenticate` doesn't require an `authToken` so it is unlikely to return `407`
`CreateLogin` uses an `authToken` that comes from a successful call to `Authenticate` and is also unlikely to return `407`, but in the off chance that it did. We'd probably want to handle the original error/response.

So I think the best move here is to return the original error or response for a `doNotRetry` for now.

We could also try to clean up `Push_Authenticate` so that a `407` will produce an `authToken` but I might want to investigate if that is a safe change some more.

Reply:
https://github.com/Expensify/Expensify.cash/pull/1187/commits/cdc77fcb7272e4426e576c3b250b035866557095

Reply:
Thanks for those thoughts. I like having `doNotRetry` for the authenticate methods because it prevents infinite loops from occurring if for some reason there is a 407 returned from the authenticate call, but I'm totally open to that not being a real problem... it just helps me sleep at night.

The resolve solution you added should do the trick!